python-package/lightgbm/basic.py

```diff
@@ -351,6 +354,60 @@ def _is_2d_collection(data: Any) -> bool:
     )
 
 
+def _is_pyarrow_table(data: Any) -> bool:
+    """Check whether data is a PyArrow table."""
+    return isinstance(data, pa_Table)
+
+
+class _ArrowCArray:
+    """Simple wrapper around the C representation of an Arrow type."""
+
+    n_chunks: int
+    chunks: arrow_cffi.CData
+    schema: arrow_cffi.CData
+
+    def __init__(self, n_chunks: int, chunks: arrow_cffi.CData, schema: arrow_cffi.CData):
+        self.n_chunks = n_chunks
+        self.chunks = chunks
+        self.schema = schema
+
+    @property
+    def chunks_ptr(self) -> int:
+        """Returns the address of the pointer to the list of chunks making up the array."""
+        return int(arrow_cffi.cast("uintptr_t", arrow_cffi.addressof(self.chunks[0])))
+
+    @property
+    def schema_ptr(self) -> int:
+        """Returns the address of the pointer to the schema of the array."""
+        return int(arrow_cffi.cast("uintptr_t", self.schema))
+
+
+@contextmanager
+def _export_arrow_to_c(data: pa_Table) -> Iterator[_ArrowCArray]:
+    """Export an Arrow type to its C representation."""
+    # Obtain objects to export
+    if isinstance(data, pa_Table):
+        export_objects = data.to_batches()
+    else:
+        raise ValueError(f"data of type '{type(data)}' cannot be exported to Arrow")
+
+    # Prepare export
+    chunks = arrow_cffi.new("struct ArrowArray[]", len(export_objects))
+    schema = arrow_cffi.new("struct ArrowSchema*")
+
+    # Export all objects
+    for i, obj in enumerate(export_objects):
+        chunk_ptr = int(arrow_cffi.cast("uintptr_t", arrow_cffi.addressof(chunks[i])))
+        if i == 0:
+            schema_ptr = int(arrow_cffi.cast("uintptr_t", schema))
+            obj._export_to_c(chunk_ptr, schema_ptr)
+        else:
+            obj._export_to_c(chunk_ptr)
+
+    yield _ArrowCArray(len(chunks), chunks, schema)
```

Comment:
What's the purpose of having this decorated with `@contextlib.contextmanager` + `yield`-ing? What's the benefit of that over just having a regular function with `return`?

Since this is only going to `yield` exactly once and doesn't have any customized setup (`__enter__()`) or teardown (`__exit__()`) behavior, I don't see the benefit of adding this complexity (e.g. an extra import, a decorator to reason about, a type hint that has to involve `Iterator`).

But I'm also not that familiar with `@contextlib.contextmanager` so I acknowledge there's probably something I'm missing.

Reply:
Fair point. My original idea was to make the scope of the C structs clear since `cffi` releases them once the variable referencing them goes out of scope. On second thought, a `with` block doesn't even help with that so let's just use a `return`.